File: app/channels/questions_channel.rb
class QuestionsChannel < ApplicationCable::Channel
  def subscribed
    stream_from "questions_channel"
  end

  def unsubscribed
    # Any cleanup needed when channel is unsubscribed
  end

  def render_question(data)
    # Use the current_user instance variable getter
    # which we set during connection in
    # app/channels/application_cable/connection.rb
    current_user = self.current_user
    current_user = nil if current_user == 'guest'

    msg = { 
      type: 'rendered question', 
      question: ApplicationController.render_with_signed_in_user(
                  current_user,
                  partial: 'questions/question',
                  locals: { question: Question.find(data['id'])}
                )
    }

    transmit(msg)
  end

end

File: app/javascript/channels/questions_channel.js
import consumer from "./consumer"

consumer.subscriptions.create("QuestionsChannel", {
  connected() {
    // Called when the subscription is ready for use on the server
  },

  disconnected() {
    // Called when the subscription has been terminated by the server
  },

  received(data) {
    if($('.questions-list').length == 0) {
      return;
    }

    // Called when there's incoming data on the websocket for this channel
    if(data['type'] == 'new question') {
      this.perform('render_question', data['question']);
    } else if (data['type'] == 'rendered question') {
      $('.questions-list').append(data['question']);
    }
  }
});
